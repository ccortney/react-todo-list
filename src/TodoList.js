import { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import {v4 as uuid} from "uuid";


const TodoList = () => {
    const [todos, setTodos] = useState([])

    console.log(todos)
    const addTodo = (newTodo) => {
        setTodos(todos => [...todos, {...newTodo, id: uuid()}])
    }

    const deleteTodo = (id) => {
        setTodos(todos => todos.filter(todo => todo.id !== id).map(todo => todo))
    }

    return (
        <div>
            <h1>Add a Todo:</h1>
            <NewTodoForm addTodo={addTodo}/>
            <h1>My Todos</h1>
            {todos.map(({id, todo, completed}) => <Todo 
                id={id}
                todo={todo} 
                completed={completed} 
                key={id}
                deleteTodo={deleteTodo}/>)}
        </div>
        
    )
}

export default TodoList;