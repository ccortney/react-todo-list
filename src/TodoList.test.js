import { render, fireEvent } from '@testing-library/react';
import TodoList from "./TodoList"

function addTodo(todoList, todo="Water Plants") {
    const input = todoList.getByLabelText("New Todo:");
    fireEvent.change(input, { target: {value: todo}});
    const button = todoList.getByText("Add Todo");
    fireEvent.click(button);
}

// smoke test
test('renders without crashing', () => {
  render(<TodoList />);
});

// snapshop test
test ('matches snapshot', () => {
  const {asFragment} = render(<TodoList/>);
  expect(asFragment()).toMatchSnapshot();
})

// more tests
test ('can add a todo', () => {
    const todoList = render(<TodoList/>)

    // todo should not exist
    expect(todoList.queryByText("Water Plants")).not.toBeInTheDocument()

    // add Todo
    addTodo(todoList)

    // todo should exist
    expect(todoList.queryByText("Water Plants")).toBeInTheDocument()

    // form should reset
    expect(todoList.getAllByDisplayValue("")).toHaveLength(1)
})

test ('can delete a todo', () => {
    const todoList = render(<TodoList/>)
    addTodo(todoList)
    expect(todoList.queryByText("Water Plants")).toBeInTheDocument()

    // Delete todo
    const removeButton  = todoList.queryByText("X");
    fireEvent.click(removeButton);
    expect(todoList.queryByText("Water Plants")).not.toBeInTheDocument()

})