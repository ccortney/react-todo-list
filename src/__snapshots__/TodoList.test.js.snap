// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Add a Todo:
    </h1>
    <form>
      <label
        for="todo"
      >
        New Todo: 
      </label>
      <input
        id="todo"
        name="todo"
        type="text"
        value=""
      />
      <button>
        Add Todo
      </button>
    </form>
    <h1>
      My Todos
    </h1>
  </div>
</DocumentFragment>
`;
